pkg/server/handler/v1alpha1/notification.go

```diff
@@ -0,0 +1,209 @@
+				}
+			}
+
+			status[endpoint.Name] = s
+			continue
+		}
+
+		log.WithField("wfr", wfr.Name).Infof("Status code of notification for %s: %d", endpoint.Name, resp.StatusCode)
+		status[endpoint.Name] = v1alpha1.NotificationStatus{
+			Result:  v1alpha1.NotificationResultSucceeded,
+			Message: fmt.Sprintf("Status code: %d", resp.StatusCode),
+		}
+	}
+
+	// Update WorkflowRun notification status with retry.
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// Get latest WorkflowRun.
+		latest, err := handler.K8sClient.CycloneV1alpha1().WorkflowRuns(wfr.Namespace).Get(wfr.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if latest.Status.Notifications == nil {
+			latest.Status.Notifications = status
+			_, err = handler.K8sClient.CycloneV1alpha1().WorkflowRuns(wfr.Namespace).Update(latest)
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		log.WithField("name", wfr.Name).Error("Update workflowrun notification status error: ", err)
+	}
+
+	return err
+}
+
+func updatePullRequestStatus(wfr *v1alpha1.WorkflowRun) error {
+	wfrName := wfr.Name
+	if wfr.Annotations == nil {
+		return fmt.Errorf("Annotations of workflowrun %s can not be empty", wfrName)
+	}
+
+	trigger, ok := wfr.Annotations[meta.AnnotationWorkflowRunTrigger]
+	if !ok {
+		return fmt.Errorf("Trigger of workflowrun %s can not be empty", wfrName)
+	}
+
+	// Skip workflowruns whose trigger are not about pull request.
+	if trigger != string(scm.PullRequestCommentEventType) && trigger != string(scm.PullRequestEventType) {
+		return nil
+	}
+
+	labels := wfr.Labels
+	if labels == nil {
+		return fmt.Errorf("Labels of workflowrun %s can not be empty", wfrName)
+	}
+
+	var tenant, project, wfName string
+	tenant = common.NamespaceTenant(wfr.Namespace)
+	if project, ok = labels[meta.LabelProjectName]; !ok {
+		return fmt.Errorf("Fail to get project name from labels of workflowrun %s", wfrName)
+	}
+	if wfName, ok = labels[meta.LabelWorkflowName]; !ok {
+		return fmt.Errorf("Fail to get workflow name from labels of workflowrun %s", wfrName)
+	}
+
+	scm, err := getSCMSourceFromWorkflowRun(wfr)
+	if err != nil {
+		return err
+	}
+
+	event, err := getSCMEventData(wfr.Annotations)
+	if err != nil {
+		return err
+	}
+
+	recordURL, err := generateRecordURL(tenant, project, wfName, wfrName)
+	if err != nil {
+		return err
+	}
+
+	return createSCMStatus(scm, wfr.Status.Overall.Phase, recordURL, event)
+}
+
+func getSCMSourceFromWorkflowRun(wfr *v1alpha1.WorkflowRun) (*s_v1alpha1.SCMSource, error) {
+	// Parse "GIT_TOKEN" from resource parameters. ONLY support one SCM git resource.
+	var gitToken string
+	found := false
+	for _, r := range wfr.Spec.Resources {
+		for i, p := range r.Parameters {
+			if p.Name == "GIT_TOKEN" {
+				gitToken = r.Parameters[i].Value
+				found = true
+				break
+			}
+		}
+
+		if found {
+			break
+		}
+	}
+
+	if found {
+		secretRef := pod.NewSecretRefValue()
```

Comment:
I'd prefer to move the ` NewSecretRefValue() ` functions to a common or util folder.

Reply:
Agree

Reply:
Found there are many dependency of this function, will do this later.